src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,372 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.BlockStore
+{
+    /// <summary>
+    /// Saves blocks to the database in batches, removes reorged blocks from the database.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The batch is saved when total serialized size of all blocks in a batch reaches <see cref="BatchThresholdSizeBytes"/>,
+    /// or when more than <see cref="BatchMaxSaveIntervalSeconds"/> passed since last batch was saved, or when node is shutting down.
+    /// </para>
+    /// <para>
+    /// When we save new blocks to the database, in case <see cref="IBlockRepository"/> contains blocks that
+    /// are no longer a part of our best chain, they are removed from the database.
+    /// </para>
+    /// <para>
+    /// Block store is being initialized we delete blocks that are not on the best chain.
+    /// </para>
+    /// </remarks>
+    public class BlockStore : IDisposable
+    {
+        /// <summary>Maximum interval between saving batches.</summary>
+        private const int BatchMaxSaveIntervalSeconds = 37;
+
+        /// <summary>Maximum number of bytes the batch can hold until the downloaded blocks are stored to the disk.</summary>
+        internal const int BatchThresholdSizeBytes = 5 * 1000 * 1000;
+
+        /// <summary>The current batch size in bytes.</summary>
+        private int currentBatchSizeBytes;
```

Comment:
@dangershony This is prime example of a needless summary that causes a maintenance overhead for no reason at all. 

If being really picky, I would rename the variable to currentBatchSizeInBytes.  Feels a bit silly to have a long variable name that is almost exactly what you mean but just skipping "In" from the name and then having a redundant summary which only brings the word "in" to the party.

Reply:
Please don't discuss things that are being discussed elsewhere on the PR review.

Reply:
I don't understand what you mean,  this is perfectly reasonable comment, we had discussions in a meeting yesterday about this where we were all present, I think it is important to highlight things that are a maintenance overhead so that we can adjust the policy to reduce this maintenance overhead.

Reply:
I think this is exactly why we have policies- to avoid such discussions during PR reviews where it's not clear what choice should be made. 
Also that's the reason why policies should be more strict.

Reply:
@majikandy No it is waste of time, please don't do that. As a reviewer your job is to follow the policies and enforce them against the code that does not follow them. Your job is not to question the policies on PR reviews. Otherwise we would have the same discussion every PR just wasting our time.

Reply:
Yep fair point, should have just messaged the example rather than on the PR

Reply:
This is a stupid discussion @Aprogiena you could just ignore it and @majikandy you could just propose a better comment.

And @majikandy you would have a better argument in the policy issue if the code was pushed.

I am gonna delete this comment thread is useless for the review